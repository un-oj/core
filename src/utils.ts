export { version } from '../package.json' with { type: 'json' };

/** Get the first key of an object. */
export function getFirstKey<T extends Record<never, never>>(obj: T): keyof T {
  return Object.keys(obj)[0] as keyof T;
}

/**
 * Parses a human-readable time string.
 * @param s The string to parse.
 * @returns The time in milliseconds, or undefined if failed.
 */
export function parseTime(s: string): number | undefined {
  const val = Number.parseInt(s);
  if (Number.isNaN(val))
    return;
  switch (s) {
    case `${val} second`:
    case `${val} seconds`:
    case `${val} sec`:
    case `${val} secs`:
      return val * 1000;
  }
}

const MEMORY_UNITS: Record<string, number> = {
  megabyte: 1024 * 1024,
  megabytes: 1024 * 1024,
  MB: 1024 * 1024,
  MiB: 1024 * 1024,
};

/**
 * Parses a human-readable memory size string.
 *
 * @param s The string to parse.
 * @returns The memory size in bytes, or undefined if failed.
 */
export function parseMemory(s: string): number | undefined {
  const [, val, unit] = s.match(/^(\d+) ([a-z]+)$/i) ?? [];
  if (!val || !unit)
    return;
  return Number(val) * MEMORY_UNITS[unit];
}

/**
 * Adds key-value pairs to the provided headers object if they aren't present.
 *
 * @param h The original headers object.
 * @param add The key-value pairs to add.
 * @returns A **new** headers object.
 */
export function addHeaders(h: HeadersInit | undefined, add: Array<[string, string]>): Record<string, string> {
  if (h instanceof Headers)
    h = h.toJSON();
  if (Array.isArray(h))
    h = Object.fromEntries(h);
  else if (h == null)
    h = {};
  else
    h = structuredClone(h);

  for (const item of add) {
    if (!(item[0] in h))
      h[item[0]] = item[1];
  }
  return h;
}

/** General error class for UnOJ. */
export class UnOJError extends Error {}
