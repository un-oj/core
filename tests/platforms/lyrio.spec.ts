import { NotFoundError } from '@un-oj/core';
import Lyrio from '@un-oj/core/platforms/lyrio';
import { describe, expect, it } from 'bun:test';
import { assertProblem } from './utils.ts';

describe('Lyrio platform (problem)', () => {
  const loj = new Lyrio();

  it('should work', async () => {
    assertProblem(await loj.getProblem('1'));
    assertProblem(await loj.getProblem('2'));
  });

  it('should throw NotFoundError w/ non-existent problem', async () => {
    expect(loj.getProblem('999999')).rejects.toThrow(NotFoundError);
  });
});
