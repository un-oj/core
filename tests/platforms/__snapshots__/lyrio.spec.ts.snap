// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Lyrio platform (problem) should work 1`] = `
{
  "description": "[{"sectionTitle":"题目描述","type":"Text","text":"输入 $ a $ 和 $ b $，输出 $ a + b $ 的结果。"},{"sectionTitle":"输入格式","type":"Text","text":"一行两个正整数 $ a $ 和 $ b $。"},{"sectionTitle":"输出格式","type":"Text","text":"一行一个正整数 $ a + b $。"},{"sectionTitle":"样例","type":"Sample","sampleId":0,"text":"根据数学知识有 $ 1 + 2 = 3 $。"},{"sectionTitle":"数据范围","type":"Text","text":"对于 $ 100\\\\% $ 的数据，$ 1 \\\\leq a, b \\\\leq 10 ^ 6 $。"}]",
  "difficulty": undefined,
  "id": "1",
  "link": "https://loj.ac/p/1",
  "memoryLimit": 536870912,
  "samples": [
    {
      "input": "1 2",
      "output": "3",
    },
  ],
  "tags": [
    {
      "color": "black",
      "id": 1,
      "name": "系统测试",
      "nameLocale": "zh_CN",
    },
  ],
  "timeLimit": 2000,
  "title": "A + B 问题",
  "type": "Traditional",
}
`;

exports[`Lyrio platform (problem) should work 2`] = `
{
  "description": "[{"sectionTitle":"题目描述","type":"Text","text":"输出 \`Hello, World!\`，大小写不限。"},{"sectionTitle":"样例 1","type":"Sample","sampleId":0,"text":""},{"sectionTitle":"样例 2","type":"Sample","sampleId":1,"text":""}]",
  "difficulty": undefined,
  "id": "2",
  "link": "https://loj.ac/p/2",
  "memoryLimit": 268435456,
  "samples": [
    {
      "input": "",
      "output": "Hello, World!",
    },
    {
      "input": "",
      "output": "hello, WORLD!",
    },
  ],
  "tags": [
    {
      "color": "black",
      "id": 1,
      "name": "系统测试",
      "nameLocale": "zh_CN",
    },
    {
      "color": "violet",
      "id": 4,
      "name": "Special Judge",
      "nameLocale": "zh_CN",
    },
  ],
  "timeLimit": 1000,
  "title": "你好，世界！",
  "type": "Traditional",
}
`;
